ONSdigital/census-rm-print-file-service
# First version of message listener

## Motivation and context
Add the component to listen to the Action.Printer queue (expecting JSON message format) and build print file rows from the incoming messages

## What has changed
* Add message listener
* Add print file templates for the 6 current IC/QM action types
* Append print file rows to partial file
* Stub file sender
* Logic to run the two components as daemon processes
* Unit tests
* Docker file, travis build, makefile
* Update README

## How to test
This service will not work with the action scheduler until it is sending JSON messages to the Action.Printer queue so for the moment you can just test it manually.
1. Run `make up` in this repo to start the print file service and rabbit
1. Navigate to the rabbit console at http://localhost:49672
1. Manually publish an action rule message in this form:
```json
{
        "actionType": "ICL1E",
        "batchId": "1",
        "batchQuantity": 3,
        "uac": "test_uac",
        "caseRef": "test_caseref",
        "title": "Mr",
        "forename": "Test",
        "surname": "McTest",
        "addressLine1": "123 Fake Street",
        "addressLine2": "Duffryn",
        "townName": "Newport",
        "postcode": "NPXXXX",
        "packCode": "P_IC_ICL1"
    }
```
1. Connect to the print file service container with `docker exec -it dev-print-file-service /bin/sh`
1. Check the contents of the `/app/partial_files` directory, the service should have created a file with the name structure `<actionType>.<packCode>.<batchId>.<batchQuantity>`
1. Check the contents of the file match the json you published to rabbit and the format matches the print file template for the action type you used
1. If you send an invalid action type the message should nack and the service will log errors (it will spin around retrying the message until you shut the app down).

## Links
https://trello.com/c/Asx8duo8/906-add-queue-consumption-to-print-file-service-8
